prov/util/src/rocr_mem_monitor.c

```diff
@@ -0,0 +1,378 @@
+
+	/* Register a deallocation callback for this ROCR memory region. */
+	hsa_ret = ofi_hsa_amd_register_deallocation_callback(hsa_info.agentBaseAddress,
+							     rocr_mm_dealloc_cb,
+							     (void *) hsa_info.sizeInBytes);
+	if (hsa_ret != HSA_STATUS_SUCCESS) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"Failed to perform hsa_amd_register_deallocation_callback: %s\n",
+			ofi_hsa_status_to_string(hsa_ret));
+
+		ret = -FI_EIO;
+		goto err_rbmap_delete_entry;
+	}
+
+	dlist_insert_tail(&(*entry)->entry, &rocr_mm.entry_list);
+
+	FI_DBG(&core_prov, FI_LOG_MR,
+	       "ROCR buffer address %p length %lu monitor entry allocated\n",
+	       hsa_info.agentBaseAddress, hsa_info.sizeInBytes);
+
+	return FI_SUCCESS;
+
+err_rbmap_delete_entry:
+	ofi_rbmap_delete(&rocr_mm.entries, (*entry)->node);
+err_free_entry:
+	free(*entry);
+err:
+	*entry = NULL;
+	return ret;
+}
+
+static void rocr_mm_init(struct ofi_mem_monitor *monitor)
+{
+	dlist_init(&rocr_mm.entry_list);
+	ofi_rbmap_init(&rocr_mm.entries, rocr_rbmap_compare);
+	ofi_monitor_init(&rocr_mm.mm);
+}
+
+static void rocr_mm_cleanup(struct ofi_mem_monitor *monitor)
+{
+	assert(dlist_empty(&rocr_mm.entry_list));
+	ofi_monitor_cleanup(&rocr_mm.mm);
+}
+
+static int rocr_mm_start(struct ofi_mem_monitor *monitor)
+{
+	/* no-op */
+	return FI_SUCCESS;
+}
+
+static void rocr_mm_stop(struct ofi_mem_monitor *monitor)
+{
+	struct rocr_mm_entry *entry;
+
+	while (!dlist_empty(&rocr_mm.entry_list)) {
+		dlist_pop_front(&rocr_mm.entry_list, struct rocr_mm_entry,
+				entry, entry);
+		rocr_mm_entry_free(entry);
+	}
+}
+
+static void rocr_mm_unsubscribe(struct ofi_mem_monitor *monitor,
+				const void *addr, size_t len,
+				union ofi_mr_hmem_info *hmem_info)
+{
+	struct rocr_mm_entry *entry;
+
+	entry = rocr_mm_entry_find(addr, len);
+	if (entry) {
+		rocr_mm_entry_free(entry);
+
+		FI_DBG(&core_prov, FI_LOG_MR,
+		       "ROCR buffer address %p length %lu unsubscribed\n",
+		       addr, len);
+	}
+}
+
+/* Subscribe is designed to monitor an entire ROCR memory region even if the
+ * user provided region is smaller. All ROCR memory regions are inserted into
+ * an RB tree for tracking. Future subscriptions will always reuse RB tree
+ * entries if possible.
+ *
+ * RB tree entries can be removed in two different ways:
+ * 1. An unsubscribe against the memory region occurs. This will occur when ROCR
+ *    invokes the deregistration callback.
+ * 2. The ROCR memory monitor is stopped.
+ *
+ * Note: The ROCR memory monitor does not impose a limit on the number of ROCR
+ * memory regions which can be monitored.
+ */
+static int rocr_mm_subscribe(struct ofi_mem_monitor *monitor, const void *addr,
+			     size_t len, union ofi_mr_hmem_info *hmem_info)
+{
+	struct rocr_mm_entry *entry;
+	int ret = FI_SUCCESS;
+
+	entry = rocr_mm_entry_find(addr, len);
+	if (entry)
+		FI_DBG(&core_prov, FI_LOG_MR,
+		       "Reusing monitored ROCR buffer address %p length %lu\n",
+		       entry->iov.iov_base, entry->iov.iov_len);
```

Comment:
I don't believe this will work.  You need some sort of reference counting on the device region, unless I'm not understanding how device regions correspond to address ranges specified by the app.

Reply:
If memory is not registered for the entire device region, this can occur. For example, if a 16K device buffer is allocate but only mapped in 4K chunks, each chunk would reuse the first chunk ROCR MM entry.

Granted a good application should register the entire buffer. When implementing rxm without FI_MR_HMEM, the internal rxm registration could hit this.

Reply:
The main reason we're in the MR cache is because the application didn't register anything and it's being done for them.  There's no reason to cache registrations initiated by the application, or to monitor them for correctness.